common/src/main/java/com/gregtechceu/gtceu/api/item/GTToolItem.java

```diff
@@ -102,16 +114,136 @@ public InteractionResult onItemUseFirst(ItemStack itemStack, UseOnContext contex
         return InteractionResult.PASS;
     }
 
+    @Override
+    public InteractionResult useOn(UseOnContext context) {
+        if (this.toolType == GTToolType.SHOVEL) {
+            return useShovelOn(context);
+        } else if (this.toolType == GTToolType.AXE) {
+            return useAxeOn(context);
+        } else if (this.toolType == GTToolType.HOE) {
+            return useHoeOn(context);
+        }
+        return InteractionResult.PASS;
+    }
+
+    public InteractionResult useShovelOn(UseOnContext context) {
+        Level level = context.getLevel();
+        BlockPos blockPos = context.getClickedPos();
+        BlockState blockState = level.getBlockState(blockPos);
+        if (context.getClickedFace() == Direction.DOWN) {
+            return InteractionResult.PASS;
+        } else {
+            Player player = context.getPlayer();
+            BlockState blockState2 = ShovelItem.FLATTENABLES.get(blockState.getBlock());
+            BlockState blockState3 = null;
+            if (blockState2 != null && level.getBlockState(blockPos.above()).isAir()) {
+                level.playSound(player, blockPos, SoundEvents.SHOVEL_FLATTEN, SoundSource.BLOCKS, 1.0F, 1.0F);
+                blockState3 = blockState2;
+            } else if (blockState.getBlock() instanceof CampfireBlock && blockState.getValue(CampfireBlock.LIT)) {
+                if (!level.isClientSide()) {
+                    level.levelEvent(null, 1009, blockPos, 0);
```

Comment:
In case `1009` is declared as a constant somewhere, it would be better to use here than a magic value.  
Otherwise, I think the context should be enough to know what it is.